core/trino-main/src/main/java/io/trino/sql/planner/JsonPathEvaluator.java

```diff
@@ -0,0 +1,1187 @@
+import static io.trino.operator.scalar.MathFunctions.absInteger;
+import static io.trino.operator.scalar.MathFunctions.absSmallint;
+import static io.trino.operator.scalar.MathFunctions.absTinyint;
+import static io.trino.operator.scalar.MathFunctions.ceilingFloat;
+import static io.trino.operator.scalar.MathFunctions.floorFloat;
+import static io.trino.spi.StandardErrorCode.PATH_EVALUATION_ERROR;
+import static io.trino.spi.type.BigintType.BIGINT;
+import static io.trino.spi.type.BooleanType.BOOLEAN;
+import static io.trino.spi.type.DateType.DATE;
+import static io.trino.spi.type.Decimals.longTenToNth;
+import static io.trino.spi.type.DoubleType.DOUBLE;
+import static io.trino.spi.type.IntegerType.INTEGER;
+import static io.trino.spi.type.RealType.REAL;
+import static io.trino.spi.type.SmallintType.SMALLINT;
+import static io.trino.spi.type.TinyintType.TINYINT;
+import static io.trino.sql.planner.JsonPathInvocationContext.ResolvedOperatorAndCoercions.RESOLUTION_ERROR;
+import static io.trino.type.DecimalCasts.longDecimalToDouble;
+import static io.trino.type.DecimalCasts.shortDecimalToDouble;
+import static java.lang.Float.floatToRawIntBits;
+import static java.lang.Float.intBitsToFloat;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Evaluates the JSON path expression using given JSON input and parameters,
+ * respecting the path mode `strict` or `lax`.
+ * Successful evaluation results in a sequence of objects.
+ * Each object in the sequence is either a `JsonNode` or a `TypedValue`
+ * Certain error conditions might be suppressed in `lax` mode.
+ * Any unsuppressed error condition causes evaluation failure.
+ * In such case, `PathEvaluationError` is thrown.
+ */
+public class JsonPathEvaluator
+{
+    private final JsonNode input;
+    private final Object[] parameters;
+    private final Session session;
+    private final ConnectorSession connectorSession;
+    private final InterpretedFunctionInvoker functionInvoker;
+    private final JsonPredicateEvaluator predicateEvaluator;
+    private final JsonPathInvocationContext invocationContext;
+    private int objectId;
+
+    public JsonPathEvaluator(JsonNode input, Object[] parameters, FunctionManager functionManager, ConnectorSession connectorSession, JsonPathInvocationContext invocationContext)
+    {
+        this.input = requireNonNull(input, "input is null");
+        this.parameters = requireNonNull(parameters, "parameters is null");
+        this.connectorSession = requireNonNull(connectorSession, "connectorSession is null");
+        this.session = ((FullConnectorSession) connectorSession).getSession();
+        this.functionInvoker = new InterpretedFunctionInvoker(functionManager);
+        this.predicateEvaluator = new JsonPredicateEvaluator(this);
+        this.invocationContext = requireNonNull(invocationContext, "invocationContext is null");
+    }
+
+    public List<Object> evaluate(IrJsonPath path)
```

Comment:
This seems backward. The fixed part should be the path (i.e., the "program" or "IR"), the functions, etc. The variable part should be the json over which the path is being evaluated.

That will allow caching resolutions and other similar decisions and attach them to the corresponding IR nodes.

Reply:
refactored